File: doubly_linked_list.rb
class Node
  attr_accessor :val, :next, :prev

  def initialize(val)
    @val = val
  end 
end


class DoublyLinkedList
  attr_accessor :head, :tail, :size

  def initialize(head = nil, tail = nil)
    @head = @tail = head
    @size = @head.nil? ? 0 : 1
  end

  def prepend(val)
    new_node = Node.new(val)
    if head.nil?
      self.head = self.tail = new_node
      @size += 1
      return
    end
    
    new_node.next = @head
    self.head.prev = new_node
    self.head = new_node

    @size += 1
  end

  def insert_at(index, val)
    return "Invalid index" if index < 0 || index > @size

    if index == 0
      return self.prepend(val)
    end
    
    if index == @size
      return self.append(val)
    end
    
    new_node = Node.new(val)
    
    # walk thru the node and setting the current pointer 
    # to set the next new node
    current = @head
    current_idx = 0
    while(current_idx < index - 1) do
      current = current.next
      current_idx += 1
    end
    
    # Before:
    # 
    # current -> current.next
    # 
    # After: 
    # 
    # current -> new_node -> current.next
    #          ^            ^
    #          |            | 
    #         prev          prev
    new_node.next = current&.next
    new_node.prev = current
    current&.next&.prev = new_node
    current&.next = new_node
      
    @size += 1
  end

  def append(val)
    new_node = Node.new(val)
    if @head.nil?
      self.head = self.tail = new_node
      return
    end

    @tail.next = new_node
    new_node.prev = @tail
    self.tail = new_node
    @size += 1
  end


  def get_by(index)
    return "Invalid index" if index < 0 || index >= @size

    if index == 0
      return @head.val
    end

    if index == @size - 1
      return @tail.val
    end 

    current = @head
    current_idx = 0
    while(current_idx < index - 1) do
      current = current.next
    end
    
    if current.next
      return current.next.val
    end
  end

  def remove(item)  
    if item == @head.val
      self.head = @head.next
      @size -= 1
      return 
    end

    if item == @tail.val
      self.tail = @tail.prev
      @size -= 1
      return
    end

    current = @head
    while(current && current.val != item)    
      current = current.next 
    end

    # current -> current.next -> current.next.next
    current&.next = current&.next&.next
    current&.prev = current&.next&.prev
    @size -= 1
  end
  
  def remove_at(index)
    return "Invalid index" if index < 0 || index >= size

    if index == 0 
      self.head = @head.next
      @size -= 1
      return
    end

    if index == @size - 1
      self.tail = @tail.prev
      @size -= 1
      return
    end

    current = @head
    curr_index = 0
    while(curr_index < index - 1) do
      current = current.next 
    end

    # current -> current.next -> current.next.next
    current&.next = current&.next&.next
    current&.prev = current&.next&.prev
    @size -= 1
  end

end

File: kata-machine/src/day1/DoublyLinkedList.ts
class Node<T> {
    value: T;
    next: Node<T> | null = null;
    prev: Node<T> | null = null;

    constructor(value: T) {
        this.value = value;
    }
}


export default class DoublyLinkedList<T> {
    private head: Node<T> | null = null;
    private tail: Node<T> | null = null;
    public length: number = 0;

    constructor() {}

    prepend(item: T): void {
        const newNode = new Node(item);
        if (!this.head) {
            this.head = this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }
        this.length++;
    }

    insertAt(item: T, idx: number): void {
        if (idx < 0 || idx > this.length) throw new Error("Invalid index");
        if (idx === 0) return this.prepend(item);
        if (idx === this.length) return this.append(item);

        const newNode = new Node(item);
        let current = this.head;
        for (let i = 0; i < idx - 1; i++) {
            if (!current) throw new Error("Invalid state");
            current = current.next;
        }
        newNode.next = current!.next;
        newNode.prev = current;
        current!.next!.prev = newNode;
        current!.next = newNode;

        this.length++;
    }

    append(item: T): void {
        const newNode = new Node(item);
        if (!this.tail) {
            this.head = this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.length++;
    }

    remove(item: T): T | undefined {
        let current = this.head;
        while (current) {
            if (current.value === item) {
                if (current.prev) current.prev.next = current.next;
                else this.head = current.next;

                if (current.next) current.next.prev = current.prev;
                else this.tail = current.prev;

                this.length--;
                return current.value;
            }
            current = current.next;
        }
        return undefined;
    }

    get(idx: number): T | undefined {
        if (idx < 0 || idx >= this.length) return undefined;
        let current = this.head;
        for (let i = 0; i < idx; i++) {
            if (!current) return undefined;
            current = current.next;
        }
        return current!.value;
    }

    removeAt(idx: number): T | undefined {
        if (idx < 0 || idx >= this.length) return undefined;
        let current = this.head;
        for (let i = 0; i < idx; i++) {
            if (!current) return undefined;
            current = current.next;
        }

        if (current!.prev) current!.prev.next = current!.next;
        else this.head = current!.next;

        if (current!.next) current!.next.prev = current!.prev;
        else this.tail = current!.prev;

        this.length--;
        return current!.value;
    }
}
